### PR TITLE
Fix/CGI.escape Function Issue

### PR DESCRIPTION
**Pull request by camlau-pp-dev**
# Overview

- Previously there was a change to the `URI.escape` function to utilize `CGI.escape` as shown in this Pull Request: https://github.com/paypal/paypalhttp_ruby/pull/19. 
- There was a breaking case with this change, and this PR serves to resolve this.
    - Links such as `https://localhost:3001/` were having the special characters be encoded, resulting in `http%3A%2F%2Flocalhost%3A3001%2F` which is incorrect. 
- This change also reverts the functionality for encoding spaces, as CGI would make " " => "+" while this URI Parser replaces spaces with " " => "%20"
----
# Testing

- Updated unit tests and added a new one to showcase the passing functionality:
RUBYHTTP_PR_IMAGE

- Integration tested this change with the [PayPal Checkout-Ruby-SDK](https://github.com/paypal/Checkout-Ruby-SDK) with Ruby version 3.0.4 and displayed no failures in unit tests.
    - Manually entered in test strings in the access token request body: https://github.com/paypal/Checkout-Ruby-SDK/blob/develop/lib/core/token_requests.rb#L7-L9
    - Temporarily added a print statement in the FormEncoded.rb file to view the results of the CGI.escape workings.

RUBY_PR_IMAGE

### CODE REVIEW

**Comment by hlahlou-pp-dev** [comment]
This change also reverts the functionality for encoding spaces, as CGI would make " " => "+" while this URI Parser replaces spaces with " " => "%20"

File: spec/paypalhttp/encoder_spec.rb [hunk 1]
```diff
@@ -116,13 +116,46 @@
...
           :another_key => 1013,
         }
       })
       serialized = Encoder.new.serialize_request(req)
 
-      expect(serialized).to eq("key=+..%3C..%3E..%25..%7B..%7D..%7C..%2F..%5E..%60..%21&another_key=1013")
+      expect(serialized).to eq("key=%20..%3C..%3E..%25..%7B..%7D..%7C..%5E..%60&another_key=1013")
+    end
+
+    it 'encodes different special/unsafe characters when using CGI.escape' do
```
Comment (hlahlou-pp-dev) [inline spec/paypalhttp/encoder_spec.rb:128]: This seems like a duplicate of the above test case. Make sure to also change the test name from mentioning CGI to mentioning URI Parser

File: spec/paypalhttp/encoder_spec.rb [hunk 2]
```diff
@@ -116,13 +116,46 @@
           "content-type" => "application/x-www-form-urlencoded; charset=utf8"
         },
         :body => {
-          :key => " ..<..>..%..{..}..|../..^..`..!",
+          :key => " ..<..>..%..{..}..|..^..`",
```
Comment (hlahlou-pp-dev) [inline spec/paypalhttp/encoder_spec.rb:119]: The characters we wanted to remove from this test were ":" and "/", why was "!" removed from the character list

File: lib/paypalhttp/serializers/form_encoded.rb [hunk 3]
```diff
@@ -1,11 +1,12 @@
-require 'cgi'
+require 'uri'
 
 module PayPalHttp
   class FormEncoded
     def encode(request)
       encoded_params = []
+      parser = URI::Parser.new()
```
Comment (crookedneighbor) [inline lib/paypalhttp/serializers/form_encoded.rb:7]: I'm not a rubyist, so I don't have a lot of context on this, but is creating a new parser for each time `encode` is called the correct thing to do? Or should the module have a single parser that is re-used?

Again, I'm not sure what the best option is for ruby. This is a genuine question.

**Review by crookedneighbor — APPROVED** [review]
Does this hav

File: lib/paypalhttp/serializers/form_encoded.rb (hunk 3 above)
Comment (hlahlou-pp-dev) [reply lib/paypalhttp/serializers/form_encoded.rb]: That is a good point. I am unsure of how Ruby handles the creation of this parser. To ensure best practice, we are moving the parser to be initialized within the class and then referenced

**Review by hlahlou-pp-dev — APPROVED** [review]

File: lib/paypalhttp/serializers/form_encoded.rb [hunk 4]
```diff
@@ -1,11 +1,15 @@
-require 'cgi'
+require 'uri'
 
 module PayPalHttp
   class FormEncoded
+    def initialize
+      @parser = parser = URI::Parser.new()
```
Comment (crookedneighbor) [inline lib/paypalhttp/serializers/form_encoded.rb:6]: Is the double assignment needed here? Shouldn't it be:
```suggestion
      @parser = URI::Parser.new()
```

Comment (hlahlou-pp-dev) [reply lib/paypalhttp/serializers/form_encoded.rb:6]: I agree, this must've just been a typo when moving the declaration into the initializer

**Pull request merged**